contract.sol

```diff
@@ -0,0 +1,152 @@
+    event PenaltiesWithdrawn(address indexed owner, uint256 amount);
+
+    // Custom errors for more efficient error handling
+    error InvalidAmount();  // Thrown when staking amount is 0 or negative
+    error InvalidRewardToken();  // Thrown when an invalid reward token is specified
+    error WithdrawAmountExceedsStake();  // Thrown when trying to withdraw more than staked
+    error NoStakeFound();  // Thrown when trying to claim reward without a stake
+    error CannotClaimRewardYet();  // Thrown when attempting to claim reward too early
+    error NoPenaltiesToWithdraw();  // Thrown when no penalties are available for withdrawal by the owner
+
+    // Initializer function to replace constructor for upgradeable contracts
+    function initialize(address _stakingToken, address[] memory _rewardTokens) public initializer {
+        __Pausable_init();  // Initialize pausable functionality
+        stakingToken = IERC20(_stakingToken);  // Set the staking token
+        minStakingPeriod = 30;  // Set the minimum staking period
+        earlyWithdrawalPenalty = 10 * 1e18;  // Set the early withdrawal penalty percentage, scaled to 1e18
+        rewardRate = 1 * 1e18;  // Set the reward rate, scaled to 1e18
+
+        // Add reward tokens and mark them as valid
+        for (uint256 i = 0; i < _rewardTokens.length; i++) {
+            rewardTokens.push(_rewardTokens[i]);
+            isRewardToken[_rewardTokens[i]] = true;
+        }
+    }
+
+    // Function to stake tokens
+    function stake(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
+        if (amount <= 0) revert InvalidAmount();  // Ensure amount is positive
+        if (!isRewardToken[rewardToken]) revert InvalidRewardToken();  // Ensure reward token is valid
+
+        stakingToken.transferFrom(msg.sender, address(this), amount);  // Transfer staked tokens to the contract
+
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        userStake.amount += amount;  // Update the staked amount
+        userStake.startTime = block.timestamp;  // Record the start time of the stake
+
+        emit Staked(msg.sender, amount, rewardToken);  // Emit a Staked event
+    }
+
+    // Function to withdraw staked tokens
+    function withdraw(uint256 amount, address rewardToken) external nonReentrant whenNotPaused {
+        Stake storage userStake = stakes[msg.sender][rewardToken];  // Access the user's stake for this reward token
+        if (userStake.amount < amount) revert WithdrawAmountExceedsStake();  // Ensure the user has enough staked
+
+        uint256 penalty = 0;
+        // Check if the withdrawal is before the minimum staking period ends
+        if (block.timestamp < userStake.startTime + minStakingPeriod) {
+            penalty = (amount * earlyWithdrawalPenalty) / 1e20;  // Calculate penalty based on scaled value
```

Comment:
Do not use 1e20 , insteead create a constant like DENOMINATOR = 1e18 , then divide by DENOMIATOR, also update the `earlyWithdrawalPenalty` as stated in the  above comment.